src/x86_64/flush.c

```diff
@@ -2,54 +2,226 @@
 #include <page.h>
 #include <apic.h>
 
-// in order to keep SMP_TEST conditionalized
+#define FLUSH_THRESHOLD 32
+#define MAX_FLUSH_ENTRIES 1024
+#define COMP_QUEUE_SIZE (MAX_FLUSH_ENTRIES*2)
+#define ENTRIES_SERVICE_THRESHOLD (MAX_FLUSH_ENTRIES/2)
+
 static boolean initialized = false;
-static queue flush_queue;
 static int flush_ipi;
 static heap flush_heap;
+static volatile word inval_gen;
+static queue free_flush_entries;
+static struct list entries;
+static int entries_count;
+static volatile boolean service_scheduled;
+static thunk flush_service;
+static queue flush_completion_queue;
+static struct spinlock flush_lock;
+
+static void queue_flush_service();
 
-typedef struct flush_entry {
-    struct refcount r;
-    u64 page; // or INVALID_ADDRESS for a full flush
-} *flush_entry;
+declare_closure_struct(1, 0, void, flush_complete,
+    flush_entry, f);
+
+struct flush_entry {
+    struct list l;
+    u64 gen;
+    u64 cpu_mask;
+    struct refcount ref;
+    boolean flush;
+    u64 pages[FLUSH_THRESHOLD];
+    int npages;
+    thunk completion;
+    closure_struct(flush_complete, finish);
+};
 
 static void invalidate (u64 page)
 {
-    asm volatile("invlpg (%0)" :: "r" (page) : "memory");            
+    asm volatile("invlpg (%0)" :: "r" (page) : "memory");
 }
 
-closure_function(0, 0, void, flush_handler)
+define_closure_function(1, 0, void, flush_complete, flush_entry, f)
+{
+    flush_entry f = bound(f);
+    assert(f->cpu_mask == 0);
+    queue_flush_service();
+}
+
+/* must be called with interrupts off */
+static void _flush_handler(void)
 {
-    flush_entry f = queue_peek(flush_queue);
-    if (f->page == INVALID_PHYSICAL) {
+    cpuinfo ci = current_cpu();
+    /* Each generation has at least one page, so if the gen difference is
+     * greater than FLUSH_THRESHOLD, just do a full tlb flush */
+    boolean full_flush = inval_gen - ci->inval_gen > FLUSH_THRESHOLD;
```

Comment:
Makes sense. It would be cool if there was a way to propagate full_flush backwards as well as forwards. So, for instance, when full_flush is set here, or later in the list traversal when f->flush is true, that would cause all previous entries in the list to be skipped and replaced with a flush for all other cpus, so they don't waste cycles invalidating pages when a flush is imminent. Two approaches come to mind: setting flush for all previous entries or keeping a global minimum generation number which is advanced to inval_gen whenever a full flush condition is met. Here, the full_flush assignment here would also be true if ci->inval_gen is less than this global minimum.

Reply:
This is a great idea! I considered both ways you suggested, but decided to set flush on all the previous entries in the list because it happens in a single place with the flush_lock held. Keeping track of a full flush generation number would be happening with multiple cores concurrently traversing the list and updating the gen number. I think it would be too racy and full of cache thrashing.